Initial SpanContext definition (#102)
open-telemetry/opentelemetry-specification: specification/tracing-api.md

TODO: Tracing operations. https://github.com/open-telemetry/opentelemetry-specification/issues/38

> ## SpanContext

Reviewer: consider adding this title in table of content on top of this file

> A SpanContext represents the portion of a Span which must be serialized and propagated along side of a distributed context. SpanContexts are immutable.

Reviewer: ```suggestion
A `SpanContext` represents the portion of a `Span` which must be serialized and propagated along side of a distributed context. `SpanContext`s are immutable. `SpanContext` MUST be a final (sealed) class.
```

> The OpenTelemetry SpanContext representation conforms to the [w3c TraceContext specification](https://www.w3.org/TR/trace-context/). It contains two identifiers - a`TraceId` and a `SpanId` - along with a set of common `TraceOptions` and system-specific `TraceState` values. SpanContext is represented as an interface, in order to be serializable into a wider variety of trace context wire formats. 

Reviewer: ```suggestion
The OpenTelemetry `SpanContext` representation conforms to the [w3c TraceContext specification](https://www.w3.org/TR/trace-context/). It contains two identifiers - a `TraceId` and a `SpanId` - along with a set of common `TraceOptions` and system-specific `TraceState` values. `SpanContext` is represented as an interface, in order to be serializable into a wider variety of trace context wire formats. 
```

TODO: Tracing operations. https://github.com/open-telemetry/opentelemetry-specification/issues/38

## SpanContext

> A SpanContext represents the portion of a Span which must be serialized and propagated along side of a distributed context. 

Reviewer: please mention that this is a final (sealed) class.

Author: I added that SpanContext is immutable, which is  an important quality. But I think you might mean something different. Is there a cross-language way of describing the intent of `final` in this case?

> 

Reviewer: Also it may be nice to mention `IsValid` API. This API is useful for consumers of API to check that there actually was a parent.

Author: Added!

> The OpenTelemetry SpanContext representation conforms to the [w3c TraceContext specification](https://w3c.github.io/trace-context). It contains two identifiers - a`TraceId` and a `SpanId` - along with a set of `TraceOptions` and `TraceState`.

Reviewer 2: Suggest to use the published version `https://www.w3.org/TR/trace-context/`.

SpanContext is represented as an interface, in order to be compatible with a wider variety of trace context wire formats.

`TraceId` A valid trace identifier is a 16-byte array with at least one non-zero byte.

`SpanId` A valid span identifier is an 8-byte array with at least one non-zero byte.

`TraceOptions` contain details about the trace. Unlike TraceState values, TraceOptions are present in all traces. Currently, the only TraceOption is a boolean `IsSampled` value.

> `TraceState` carries system-specific configuration data, represented as a list of key-value pairs. TraceState allows multiple tracing systems to participate in the same trace. 

Reviewer: ```suggestion
`Tracestate` carries system-specific configuration data, represented as a list of key-value pairs. TraceState allows multiple tracing systems to participate in the same trace. 
```

I believe we prefer a single word in a language. I'm fine either way

Reviewer 3: We had a long long discussion about the naming before in https://github.com/census-instrumentation/opencensus-specs/issues/168. The conclusion back then is to use `Tracestate` (as one word). My personal opinion is let's stick with the existing name...

Reviewer 2: https://github.com/open-telemetry/opentelemetry-python/pull/11#discussion_r290583447

Author: I updated this. But... it sure looks funky sitting next to TraceOptions. Related naming glitch: there is no such thing as TraceOptions in the spec. they are referred to as `trace-flags`: https://www.w3.org/TR/trace-context/#trace-flags

Do we want `Traceflags` and `Tracestate` instead of `TraceOptions` and `Tracestate`?

Reviewer: Please file an issue suggesting traceflags instead of traceoptions. Makes sense

> TraceState keys are opaque strings up to 256 characters. They MUST begin with a lowercase letter, and can only contain lowercase letters `a-z`, digits `0-9`, underscores `_`, dashes `-`, asterisks `*`, and forward slashes `/`.

Reviewer 2: Do we want to describe the syntax rules here or point people to the spec?

Reviewer 4: I think pointing to the spec would be better (and easier).

Usually this will be the W3C Trace Context as the HTTP text format. For more details, see
[trace-context](https://github.com/w3c/trace-context).

## SpanContext

> A `SpanContext` represents the portion of a `Span` which must be serialized and propagated along side of a distributed context. `SpanContext`s are immutable. `SpanContext` MUST be a final (sealed) class.

Author: @SergeyKanzhelev I added the bit about final/sealed being required. But I'm not sure that language is understandable in the context of a cross-language spec. Can you explain the goal/reasoning for sealing this class?

Reviewer: `SpanContext` should not allow alternative implementations that carry additional data. So all extensibility will be made via `tracestate` AND API users will not assume any surprises from SpanContext like keeping a references on some memory-heavy objects

Author: Ok. This extensibility issue could also be referred to as "casting" in other languages. So, we are advising that casting or extending the SpanContext interface is almost never correct.
